### PR TITLE
marshal: add marshalers, fix doc

### PR DESCRIPTION
**Pull request by drmingdrmer**
# Description

Interface changes:

-   Add `GetSize()` to interface Marshaler, for getting marshaled size
    from a unmarshaled value.

Interface implements changes:

-   Add struct `Bytes` marshaler to deal with `[]byte`.

-   Auto generate int marshalers and test with `go generate ./marshal`.
    Support int16, int32, int64, uint16, uint32, uint64.

-   Add struct TypeMarshaler to marshal any type of fixed-size.
    What is size fixed type:

        `int32`:                      fixed size
        `struct { X int32; Y byte }`: fixed size
        `int`:                        not fixed size
        `[]int32`:                    not fixed size

Other:

-   Fix doc: double `l` to single `l` in marshal[l]er, marshal[l]ed etc.

-   Fix lint issues

## Type of change

- **Bug fix**         
- **New feature**     
- **Refactoring**
- **Document changes**
- **Test changes**

# Checklist:

- [x] **Style**:       My code follows the **style guidelines** of this project
- [x] **Self-review**: I have performed a **self-review** of my own code
- [x] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [x] **Doc**:         I have made corresponding changes to the **documentation**
- [x] **No-warnings**: My changes generate **no new warnings**
- [x] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes
- [ ] **Dep**:         Any **dependent** changes have been merged and published in downstream modules

### CODE REVIEW

File: marshal/bytes.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// Marshal converts byte slice to byte slice.
+func (c Bytes) Marshal(d interface{}) []byte {
+	return d.([]byte)
+}
+
+// Unmarshal copies fixed length slice out of source byte slice.
+// The returned bytes are NOT copied.
+func (c Bytes) Unmarshal(b []byte) (int, interface{}) {
+	s := b[:c.Size]
+	return c.Size, s
```
Comment (pengsven) [inline marshal/bytes.go]: 第一个参数返回实际的len(s)如何？

File: marshal/type_marshaler.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	return m.Size, reflect.Indirect(v).Interface()
+}
+
+// GetSize returns m.Size.
+func (m *TypeMarshaler) GetSize(d interface{}) int {
+	return m.Size
+}
+
+// GetMarshaledSize returns m.Size.
+func (m *TypeMarshaler) GetMarshaledSize(b []byte) int {
```
Comment (pengsven) [inline marshal/type_marshaler.go]: 这2个函数的新参没用到，是为和Marshaler的2个函数保持一致吗

Comment (drmingdrmer) [reply marshal/type_marshaler.go]: 对.必须要实现Marshaler接口

File: marshal/bytes.go (hunk 1 above)
Comment (drmingdrmer) [reply marshal/bytes.go]: len(s)和 b.Size是一样的啊...如果不一样17行会panic

**Pull request merged**